components/eamxx/src/control/intensive_observation_period.cpp

```diff
@@ -237,6 +495,354 @@ read_fields_from_file_for_iop (const std::string& file_name,
+    const auto hybm_v = m_helper_fields["hybm"].get_view<const Real*>();
+    Kokkos::parallel_for(model_nlevs, KOKKOS_LAMBDA (const int ilev) {
+      model_pres_v(ilev) = 1000*hyam_v(ilev) + Ps()*hybm_v(ilev)/100;
+    });
+
+    // Find file pressure levels just outside the range of model pressure levels
+    Kokkos::parallel_reduce(adjusted_file_levs, KOKKOS_LAMBDA (const int& ilev, int& lmax, int& lmin) {
+      if (iop_file_pres_v(ilev) <= model_pres_v(0) && ilev > lmax) {
+        lmax = ilev;
+      }
+      if (iop_file_pres_v(ilev) >= model_pres_v(model_nlevs-1) && ilev+1 < lmin) {
+        lmin = ilev+1;
+      }
+    },
+    Kokkos::Max<int>(iop_file_start),
+    Kokkos::Min<int>(iop_file_end));
+
+    // Find model pressure levels just inside range of file pressure levels
+    Kokkos::parallel_reduce(model_nlevs, KOKKOS_LAMBDA (const int& ilev, int& lmin, int& lmax) {
+      if (model_pres_v(ilev) >= iop_file_pres_v(iop_file_start) && ilev < lmin) {
+        lmin = ilev;
+      }
+      if (model_pres_v(ilev) <= iop_file_pres_v(iop_file_end-1) && ilev+1 > lmax) {
+        lmax = ilev+1;
+      }
+    },
+    Kokkos::Min<int>(model_start),
+    Kokkos::Max<int>(model_end));
+  }
+
+  // Loop through fields and store data from file
+  for (auto& it : m_iop_fields) {
+    auto fname = it.first;
+    auto field = it.second;
+
+    // File may use different varname than IOP class
+    auto file_varname = (m_iop_file_varnames.count(fname) > 0) ? m_iop_file_varnames[fname] : fname;
+
+    if (field.rank()==0) {
+      // For scalar data, read iop file variable directly into field data
+      auto data = field.get_view<Real, Host>().data();
+      read_variable_from_file(iop_file, file_varname, "real", {"lon","lat"}, iop_file_time_idx, data);
+      field.sync_to_dev();
+    } else if (field.rank()==1) {
+      // Create temporary fields for reading iop file variables. We use
+      // adjusted_file_levels (computed above) which contains an unset
+      // value for surface.
+      FieldIdentifier fid(file_varname+"_iop_file",
+                          FieldLayout({FieldTag::LevelMidPoint},
+                          {adjusted_file_levs}),
+                          ekat::units::Units::nondimensional(),
+                          "");
+      Field iop_file_field(fid);
+      iop_file_field.allocate_view();
+
+      // Read data from iop file.
+      std::vector<Real> data(file_levs);
+      read_variable_from_file(iop_file, file_varname, "real", {"lon","lat","lev"}, iop_file_time_idx, data.data());
+
+      // Copy first adjusted_file_levs-1 values to field
+      auto iop_file_v_h = iop_file_field.get_view<Real*,Host>();
+      for (int ilev=0; ilev<adjusted_file_levs-1; ++ilev) iop_file_v_h(ilev) = data[ilev];
+
+      // Set or compute surface value
+      const auto has_srf = m_iop_field_surface_varnames.count(fname)>0;
+      if (has_srf) {
+        const auto srf_varname = m_iop_field_surface_varnames[fname];
+        Real srf_data;
```

Comment:
Nitpicky, but why can't you just pass `&iop_file_v_h(adjusted_file_levs-1)`, and do away with `srf_data`?